File: CipherCore/Sources/CipherCore/Util/Support.swift
import Foundation
import Security

public enum CipherError: Error, Equatable {
    case malformed
    case notAKryptosMessage
    case decryptionFailed
    case stegoCapacityExceeded
    case invalidInput
}

public func randomBytes(_ count: Int) -> Data {
    guard count > 0 else { return Data() }
    var bytes = [UInt8](repeating: 0, count: count)
    let status = bytes.withUnsafeMutableBytes { buffer -> Int32 in
        guard let base = buffer.baseAddress else { return errSecParam }
        return SecRandomCopyBytes(kSecRandomDefault, count, base)
    }
    guard status == errSecSuccess else {
        preconditionFailure("SecRandomCopyBytes failed with \(status)")
    }
    return Data(bytes)
}

public struct BinaryWriter {
    public private(set) var data = Data()
    public init() {}

    public mutating func writeByte(_ b: UInt8) { data.append(b) }

    public mutating func writeUInt32(_ v: UInt32) {
        data.append(UInt8((v >> 24) & 0xFF))
        data.append(UInt8((v >> 16) & 0xFF))
        data.append(UInt8((v >> 8) & 0xFF))
        data.append(UInt8(v & 0xFF))
    }

    public mutating func writeVar(_ d: Data) {
        writeUInt32(UInt32(d.count))
        data.append(d)
    }
}

public struct BinaryReader {
    private let bytes: [UInt8]
    private var index = 0

    public init(_ data: Data) { self.bytes = [UInt8](data) }

    public mutating func readByte() throws -> UInt8 {
        guard index < bytes.count else { throw CipherError.malformed }
        defer { index += 1 }
        return bytes[index]
    }

    public mutating func readUInt32() throws -> UInt32 {
        let b = try readRaw(4)
        let a = [UInt8](b)
        return (UInt32(a[0]) << 24) | (UInt32(a[1]) << 16) | (UInt32(a[2]) << 8) | UInt32(a[3])
    }

    public mutating func readRaw(_ n: Int) throws -> Data {
        guard n >= 0, n <= bytes.count - index else { throw CipherError.malformed }
        defer { index += n }
        return Data(bytes[index ..< index + n])
    }

    public mutating func readVar() throws -> Data {
        let n = try readUInt32()
        return try readRaw(Int(n))
    }
}

File: android/app/src/main/java/com/kryptos/android/core/Support.kt
package com.kryptos.android.core

import java.security.SecureRandom

class CipherException(val kind: Kind) : Exception(kind.name) {
    enum class Kind {
        MALFORMED,
        NOT_A_KRYPTOS_MESSAGE,
        DECRYPTION_FAILED,
        STEGO_CAPACITY_EXCEEDED,
        INVALID_INPUT,
        UNSUPPORTED_FORMAT,
    }
}

private val rng = SecureRandom()

fun randomBytes(count: Int): ByteArray = ByteArray(count).also { rng.nextBytes(it) }

private val HEX_DIGITS = "0123456789abcdef".toCharArray()

fun hexOf(bytes: ByteArray): String {
    val out = CharArray(bytes.size * 2)
    for (i in bytes.indices) {
        val v = bytes[i].toInt() and 0xFF
        out[i * 2] = HEX_DIGITS[v ushr 4]
        out[i * 2 + 1] = HEX_DIGITS[v and 0x0F]
    }
    return String(out)
}

fun sha256Hex(data: ByteArray): String =
    hexOf(java.security.MessageDigest.getInstance("SHA-256").digest(data))

internal object StegoTokenizer {
    fun isHan(cp: Int): Boolean =
        cp in 0x4E00..0x9FFF || cp in 0x3400..0x4DBF || cp in 0xF900..0xFAFF

    fun split(text: String): List<String> {
        val normalized = java.text.Normalizer.normalize(text, java.text.Normalizer.Form.NFC).lowercase()
        val tokens = ArrayList<String>()
        val current = StringBuilder()
        var i = 0
        while (i < normalized.length) {
            val cp = normalized.codePointAt(i)
            if (isHan(cp)) {
                if (current.isNotEmpty()) {
                    tokens.add(current.toString())
                    current.setLength(0)
                }
                tokens.add(String(Character.toChars(cp)))
            } else if (Character.isLetter(cp)) {
                current.appendCodePoint(cp)
            } else if (current.isNotEmpty()) {
                tokens.add(current.toString())
                current.setLength(0)
            }
            i += Character.charCount(cp)
        }
        if (current.isNotEmpty()) tokens.add(current.toString())
        return tokens
    }

    fun runs(text: String): List<String> {
        val normalized = java.text.Normalizer.normalize(text, java.text.Normalizer.Form.NFC).lowercase()
        val tokens = ArrayList<String>()
        val current = StringBuilder()
        var i = 0
        while (i < normalized.length) {
            val cp = normalized.codePointAt(i)
            if (Character.isLetter(cp)) {
                current.appendCodePoint(cp)
            } else if (current.isNotEmpty()) {
                tokens.add(current.toString())
                current.setLength(0)
            }
            i += Character.charCount(cp)
        }
        if (current.isNotEmpty()) tokens.add(current.toString())
        return tokens
    }
}

object CachePurge {
    private val hooks = java.util.concurrent.CopyOnWriteArrayList<() -> Unit>()
    private val decryptedHooks = java.util.concurrent.CopyOnWriteArrayList<() -> Unit>()
    fun register(hook: () -> Unit) { hooks += hook }
    fun registerDecrypted(hook: () -> Unit) { decryptedHooks += hook }
    fun purgeAll() { hooks.forEach { it() } }
    fun purgeDecrypted() { decryptedHooks.forEach { it() } }
}

internal class WipingBuffer(initial: Int = 8192) : java.io.OutputStream() {
    private var buf = ByteArray(initial.coerceAtLeast(64))
    private var size = 0

    private fun ensure(extra: Int) {
        if (extra < 0) throw CipherException(CipherException.Kind.INVALID_INPUT)
        val needed = size.toLong() + extra
        if (needed <= buf.size) return
        if (needed > MAX_CAPACITY) throw CipherException(CipherException.Kind.INVALID_INPUT)
        var capacity = buf.size.toLong()
        while (capacity < needed) capacity = minOf(capacity * 2, MAX_CAPACITY)
        val grown = buf.copyOf(capacity.toInt())
        buf.fill(0)
        buf = grown
    }

    override fun write(b: Int) {
        ensure(1)
        buf[size++] = b.toByte()
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        ensure(len)
        b.copyInto(buf, size, off, off + len)
        size += len
    }

    fun drain(): ByteArray {
        val out = buf.copyOf(size)
        buf.fill(0)
        size = 0
        return out
    }

    companion object {
        const val MAX_CAPACITY = Int.MAX_VALUE - 8L
    }
}

internal inline fun wipingBytes(write: (java.io.OutputStream) -> Unit): ByteArray {
    val buffer = WipingBuffer()
    return try {
        write(buffer)
        buffer.drain()
    } catch (e: Throwable) {
        buffer.drain().fill(0)
        throw e
    }
}

class BinaryWriter {
    private val out = java.io.ByteArrayOutputStream()
    val data: ByteArray get() = out.toByteArray()

    fun writeByte(b: Int) = out.write(b and 0xFF)

    fun writeUInt32(v: Long) {
        out.write(((v shr 24) and 0xFF).toInt())
        out.write(((v shr 16) and 0xFF).toInt())
        out.write(((v shr 8) and 0xFF).toInt())
        out.write((v and 0xFF).toInt())
    }

    fun writeVar(d: ByteArray) {
        writeUInt32(d.size.toLong())
        out.write(d)
    }
}

class BinaryReader(private val bytes: ByteArray) {
    private var index = 0

    fun readByte(): Int {
        if (index >= bytes.size) throw CipherException(CipherException.Kind.MALFORMED)
        return bytes[index++].toInt() and 0xFF
    }

    fun readUInt32(): Long {
        val b = readRaw(4)
        return ((b[0].toLong() and 0xFF) shl 24) or ((b[1].toLong() and 0xFF) shl 16) or
            ((b[2].toLong() and 0xFF) shl 8) or (b[3].toLong() and 0xFF)
    }

    fun readRaw(n: Int): ByteArray {
        if (n < 0 || n > bytes.size - index) throw CipherException(CipherException.Kind.MALFORMED)
        val r = bytes.copyOfRange(index, index + n)
        index += n
        return r
    }

    fun readVar(): ByteArray = readRaw(readUInt32().toInt())
}

class TaskQueue(name: String) {
    private val pool = java.util.concurrent.Executors.newSingleThreadExecutor { r ->
        Thread(r, name).apply { isDaemon = true }
    }

    fun execute(block: () -> Unit) {
        if (pool.isShutdown) return
        runCatching { pool.execute { runCatching(block) } }
    }

    fun shutdown() {
        runCatching { pool.shutdown() }
    }

    fun shutdownNow() {
        runCatching { pool.shutdownNow() }
    }
}
